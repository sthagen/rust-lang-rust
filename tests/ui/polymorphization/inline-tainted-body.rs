//@ compile-flags: -Zvalidate-mir -Zinline-mir=yes

#![feature(unboxed_closures)]

use std::sync::Arc;

pub struct WeakOnce<T>();
//~^ ERROR type parameter `T` is never used

impl<T> WeakOnce<T> {
    extern "rust-call" fn try_get(&self) -> Option<Arc<T>> {}
    //~^ ERROR functions with the "rust-call" ABI must take a single non-self tuple argument
    //~| ERROR mismatched types

    pub fn get(&self) -> Arc<T> {
        self.try_get()
            .unwrap_or_else(|| panic!("Singleton {} not available", std::any::type_name::<T>()))
    }
}

fn main() {}
