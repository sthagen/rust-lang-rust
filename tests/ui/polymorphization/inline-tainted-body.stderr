error[E0392]: type parameter `T` is never used
  --> $DIR/inline-tainted-body.rs:7:21
   |
LL | pub struct WeakOnce<T>();
   |                     ^ unused type parameter
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
   = help: if you intended `T` to be a const parameter, use `const T: /* Type */` instead

error: functions with the "rust-call" ABI must take a single non-self tuple argument
  --> $DIR/inline-tainted-body.rs:11:35
   |
LL |     extern "rust-call" fn try_get(&self) -> Option<Arc<T>> {}
   |                                   ^^^^^

error[E0308]: mismatched types
  --> $DIR/inline-tainted-body.rs:11:45
   |
LL |     extern "rust-call" fn try_get(&self) -> Option<Arc<T>> {}
   |                           -------           ^^^^^^^^^^^^^^ expected `Option<Arc<T>>`, found `()`
   |                           |
   |                           implicitly returns `()` as its body has no tail or `return` expression
   |
   = note:   expected enum `Option<Arc<T>>`
           found unit type `()`

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0308, E0392.
For more information about an error, try `rustc --explain E0308`.
